schedview/app/prenight/prenight.py

```diff
@@ -1,418 +1,648 @@
+            ),
+            pn.Column(
+                "<h2>Astronomical Events</h2>",
+                pn.param.ParamMethod(
+                    prenight.almanac_events_table, loading_indicator=True
+                ),
+            ),
         ),
-        "<h2>Parameters</h2>",
-        night,
-        timezone,
-        scheduler_fname,
-        opsim_output_fname,
-        "<h2>Astronomical Events</h2>",
-        pn.Row(pn.bind(almanac_events, night, timezone)),
-        "<h2>Simulated visits</h2>",
-        pn.Row(pn.bind(visit_explorer, opsim_output_fname, night)),
-        pn.Row(pn.bind(visit_table, opsim_output_fname, night)),
-        pn.Row(pn.bind(visit_skymaps, opsim_output_fname, scheduler_fname, night)),
-    )
+        pn.Tabs(
+            (
+                "Visit explorer",
+                pn.param.ParamMethod(prenight.visit_explorer, loading_indicator=True),
+            ),
+            (
+                "Table of visits",
+                pn.param.ParamMethod(prenight.visit_table, loading_indicator=True),
+            ),
+            (
+                "Sky maps",
+                pn.param.ParamMethod(prenight.visit_skymaps, loading_indicator=True),
+            ),
+            (
+                "Reward plots",
+                pn.Column(
+                    pn.param.ParamMethod(
+                        prenight.reward_params, loading_indicator=True
+                    ),
+                    pn.param.ParamMethod(prenight.reward_plot, loading_indicator=True),
+                    pn.param.ParamMethod(
+                        prenight.infeasible_plot, loading_indicator=True
+                    ),
+                ),
+            ),
+            dynamic=False,  # When true, visit_table never renders. Why?
+        ),
+        debug_info,
+    ).servable()
 
-    if reward_df is not None:
-        reward_rows = pn.Column(
-            "<h2>Reward values through the night</h2>",
-            pn.Row(tier, surveys, basis_function),
-            reward_plot,
-            infeasible_plot,
-        )
-        app = pn.Column(basic_app, reward_rows)
-    else:
-        app = basic_app
+    def clear_caches(session_context):
+        logging.info("session cleared")
+        pn_app.stop()
 
-    return app
+    try:
+        pn.state.on_session_destroyed(clear_caches)
+    except RuntimeError as e:
+        logging.info("RuntimeError: %s", e)
+
+    if return_app:
+        return pn_app, prenight
+    else:
+        return pn_app
 
 
 if __name__ == "__main__":
-    app = prenight_app()
+    print("Starting prenight dashboard")
 
-    template = pn.Template(
-        """
-        {% extends base %}
-        {% block postamble %}
-        <script src="https://unpkg.com/gpu.js@latest/dist/gpu-browser.min.js"></script>
-        {% endblock %}
-    """
+    if "PRENIGHT_PORT" in os.environ:
+        prenight_port = int(os.environ["PRENIGHT_PORT"])
+    else:
+        prenight_port = 8080
+
```

Comment:
It might be nice if this could use arguments as well as or instead of the os.environment variables, but I see that there are no arguments being passed here at all so i’m not sure if they’re just not available somehow with how this runs?

Reply:
This is a good idea, but maybe not for this ticket. I just created a new one for it: [PREOPS-3600](https://jira.lsstcorp.org/browse/PREOPS-3600).